efficientdet/keras/efficientdet_arch_keras.py

```diff
@@ -25,64 +25,239 @@
+    self.vars = []
 
-    self.fpn_config = None
-    self.fpn_name = fpn_name
+  def fuse_features(self, nodes):
+    """Fuse features from different resolutions and return a weighted sum.
 
-    super(BiFPNLayer, self).__init__(**kwargs)
+    Args:
+      nodes: a list of tensorflow features at different levels
+      weight_method: feature fusion method. One of:
+        - "attn" - Softmax weighted fusion
+        - "fastattn" - Fast normalzied feature fusion
+        - "sum" - a sum of inputs
+
+    Returns:
+      A tensor denoting the fused feature.
+    """
+    dtype = nodes[0].dtype
+
+    if self.weight_method == 'attn':
+      edge_weights = []
+      for _ in nodes:
+        var = tf.Variable(1.0, name='WSM')
+        self.vars.append(var)
+        var = tf.cast(var, dtype=dtype)
+        edge_weights.append(var)
+      normalized_weights = tf.nn.softmax(tf.stack(edge_weights))
+      nodes = tf.stack(nodes, axis=-1)
+      new_node = tf.reduce_sum(nodes * normalized_weights, -1)
+    elif self.weight_method == 'fastattn':
+      edge_weights = []
+      for _ in nodes:
+        var = tf.Variable(1.0, name='WSM')
+        self.vars.append(var)
+        var = tf.cast(var, dtype=dtype)
+        edge_weights.append(var)
+      weights_sum = tf.add_n(edge_weights)
+      nodes = [
+          nodes[i] * edge_weights[i] / (weights_sum + 0.0001)
+          for i in range(len(nodes))
+      ]
+      new_node = tf.add_n(nodes)
+    elif self.weight_method == 'channel_attn':
+      num_filters = int(nodes[0].shape[-1])
+      edge_weights = []
+      for _ in nodes:
+        var = tf.Variable(lambda: tf.ones([num_filters]), name='WSM')
+        self.vars.append(var)
+        var = tf.cast(var, dtype=dtype)
+        edge_weights.append(var)
+      normalized_weights = tf.nn.softmax(tf.stack(edge_weights, -1), axis=-1)
+      nodes = tf.stack(nodes, axis=-1)
+      new_node = tf.reduce_sum(nodes * normalized_weights, -1)
+    elif self.weight_method == 'channel_fastattn':
+      num_filters = int(nodes[0].shape[-1])
+      edge_weights = []
+      for _ in nodes:
+        var = tf.Variable(lambda: tf.ones([num_filters]), name='WSM')
+        self.vars.append(var)
+        var = tf.cast(var, dtype=dtype)
+        edge_weights.append(var)
+
+      weights_sum = tf.add_n(edge_weights)
+      nodes = [
+          nodes[i] * edge_weights[i] / (weights_sum + 0.0001)
+          for i in range(len(nodes))
+      ]
+      new_node = tf.add_n(nodes)
+    elif self.weight_method == 'sum':
+      new_node = tf.add_n(nodes)
+    else:
+      raise ValueError('unknown weight_method {}'.format(self.weight_method))
 
-  def call(self, feats):
-    # @TODO: Implement this with keras logic
-    return legacy_arch.build_bifpn_layer(feats, self.feat_sizes, self)
+    return new_node
 
-  def get_config(self):
-    base_config = super(BiFPNLayer, self).get_config()
+  def call(self, feats):
+    # num_output_connections = [0 for _ in feats]
```

Comment:
num_output_connections only for debug?

Reply:
yes, it was for debugging, and you don't need it here.